packages/user-operation-controller/src/utils/gas-fees.ts

```diff
@@ -0,0 +1,295 @@
+  getGetFasEstimates: () => Promise<SuggestedGasFees>,
+  userOperation: UserOperation,
+  transaction?: TransactionParams,
+) {
+  const { maxFeePerGas, maxPriorityFeePerGas } = originalRequest;
+  const { gasPrice } = transaction ?? {};
+  const { maxFeePerGas: updatedMaxFeePerGas } = userOperation;
+
+  if (!isGasFeeEmpty(maxPriorityFeePerGas)) {
+    log('Using maxPriorityFeePerGas from request', maxPriorityFeePerGas);
+    return maxPriorityFeePerGas as string;
+  }
+
+  if (isGasFeeEmpty(maxFeePerGas) && gasPrice) {
+    log('Setting maxPriorityFeePerGas to transaction gasPrice', gasPrice);
+    return gasPrice;
+  }
+
+  const { maxPriorityFeePerGas: suggestedMaxPriorityFeePerGas } =
+    await getGetFasEstimates();
+
+  if (suggestedMaxPriorityFeePerGas) {
+    log(
+      'Using maxPriorityFeePerGas from estimate',
+      suggestedMaxPriorityFeePerGas,
+    );
+    return suggestedMaxPriorityFeePerGas;
+  }
+
+  log('Setting maxPriorityFeePerGas to maxFeePerGas', updatedMaxFeePerGas);
+
+  return updatedMaxFeePerGas;
+}
+
+/**
+ * Gets the userFeeLevel for a user operation.
+ * @param metadata - The metadata for the user operation.
+ * @param originalRequest - The original request to add the user operation.
+ * @param suggestedGasFees - The suggested gas fees, if any.
+ * @param transaction - The transaction that created the user operation.
+ * @returns The userFeeLevel for the user operation.
+ */
+function getUserFeeLevel(
+  metadata: UserOperationMetadata,
+  originalRequest: AddUserOperationRequest,
+  suggestedGasFees?: SuggestedGasFees,
+  transaction?: TransactionParams,
+): UserFeeLevel {
+  const { origin } = metadata;
+  const { maxFeePerGas, maxPriorityFeePerGas } = originalRequest;
+
+  const {
+    maxFeePerGas: suggestedMaxFeePerGas,
+    maxPriorityFeePerGas: suggestedMaxPriorityFeePerGas,
+  } = suggestedGasFees || {};
+
+  if (
+    isGasFeeEmpty(maxFeePerGas) &&
+    isGasFeeEmpty(maxPriorityFeePerGas) &&
+    transaction?.gasPrice
+  ) {
+    return origin === ORIGIN_METAMASK
+      ? UserFeeLevel.CUSTOM
+      : UserFeeLevel.DAPP_SUGGESTED;
+  }
+
+  if (
+    isGasFeeEmpty(maxFeePerGas) &&
+    isGasFeeEmpty(maxPriorityFeePerGas) &&
+    suggestedMaxFeePerGas &&
+    suggestedMaxPriorityFeePerGas
+  ) {
+    return UserFeeLevel.MEDIUM;
+  }
+
+  if (origin === ORIGIN_METAMASK) {
+    return UserFeeLevel.CUSTOM;
+  }
+
+  return UserFeeLevel.DAPP_SUGGESTED;
+}
+
+/**
+ * Gets suggested gas fees.
+ * @param request - The request to update the gas fees.
+ * @param request.getGasFeeEstimates - A callback to get gas fee estimates.
+ * @param request.provider - A provider to query the network.
+ * @returns The suggested gas fees.
+ */
+async function getSuggestedGasFees(request: UpdateGasFeesRequest) {
+  const { getGasFeeEstimates, provider } = request;
+
+  try {
+    const { gasFeeEstimates, gasEstimateType } = await getGasFeeEstimates();
+
+    if (gasEstimateType === GAS_ESTIMATE_TYPES.FEE_MARKET) {
```

Comment:
Would a switch statement be better than this series of if statements?

Reply:
I think it could be if we were explicitly using every type with a fallback, but the flow here is more complex since we also run the "default" in the event of an error or if the initial types don't have suitable values.

Reply:
You've addressed all the values in GAS_ESTIMATE_TYPES except "NONE", which feels like a default case. 

The purpose of your code appears to be:
"To calculate the suggested gas fees for a given GAS_ESTIMATE_TYPES value."

If you plan to add another GAS_ESTIMATE_TYPES value, would you also add a corresponding if-statement? If that's the case, consider implementing a switch statement with an empty default case (possibly noting that it handles "NONE"). This approach might better reflect the intent of your code.

Also, see comment below about moving the fallback into a function.